components/ws-daemon/pkg/daemon/cgroup_customizer.go

```diff
@@ -64,8 +78,72 @@ func (c *CgroupCustomizer) WorkspaceAdded(ctx context.Context, ws *dispatch.Work
 	return nil
 }
 
-func (c *CgroupCustomizer) customV1() ([]cgroups.Subsystem, error) {
+func (c *CgroupCustomizerV1) customV1() ([]cgroups.Subsystem, error) {
 	return []cgroups.Subsystem{
 		cgroups.NewDevices(c.cgroupBasePath),
 	}, nil
 }
+
+type CgroupCustomizerV2 struct {
+	cgroupBasePath string
+}
+
+func (c *CgroupCustomizerV2) WorkspaceAdded(ctx context.Context, ws *dispatch.Workspace) error {
+	cgroupPath, err := retrieveCGroupPath(ctx, ws.ContainerID)
+	if err != nil {
+		return err
+	}
+
+	fullCgroupPath := filepath.Join(c.cgroupBasePath, cgroupPath)
+	cgroupFD, err := unix.Open(fullCgroupPath, unix.O_DIRECTORY|unix.O_RDONLY, 0o600)
+	if err != nil {
+		return xerrors.Errorf("cannot get directory fd for %s", fullCgroupPath)
+	}
+	defer unix.Close(cgroupFD)
+
+	insts, license, err := devicefilter.DeviceFilter(composeDeviceRules())
+	if err != nil {
+		return xerrors.Errorf("failed to generate device filter: %w", err)
+	}
+
+	_, err = ebpf.LoadAttachCgroupDeviceFilter(insts, license, cgroupFD)
+	return err
+}
+
+func retrieveCGroupPath(ctx context.Context, id container.ID) (string, error) {
+	disp := dispatch.GetFromContext(ctx)
+	if disp == nil {
+		return "", xerrors.Errorf("no dispatch available")
+	}
+
+	cgroupPath, err := disp.Runtime.ContainerCGroupPath(context.Background(), id)
+	if err != nil {
+		return "", xerrors.Errorf("cannot retrieve cgroup path: %w", err)
+	}
+
+	return cgroupPath, nil
+}
+
+func composeDeviceRules() []*devices.Rule {
+	denyAll := devices.Rule{
+		Type:        'a',
+		Permissions: "rwm",
+		Allow:       false,
+	}
+
+	allowFuse := devices.Rule{
+		Type:        'c',
+		Major:       fuseDeviceMajor,
+		Minor:       fuseDeviceMinor,
+		Permissions: "rwm",
+		Allow:       true,
+	}
+
+	deviceRules := make([]*devices.Rule, 0)
+	deviceRules = append(deviceRules, &denyAll, &allowFuse)
+	for _, device := range specconv.AllowedDevices {
```

Comment:
In our current use case, we don't specify cgroup devices. In other words, this is all we need to specify for now. However, if we want to add some devices about the runc layer(e.g. kubelet, contained) in the future, this part will arise a problem. However, it may be an option to decide that the cgroup devices are managed here. In fact, we have been used to overwrite `/dev/fuse` here.
@ChristinFrohne WDYT?

Reply:
What about applying it to the pod cgroup instead of the container cgroup and erasing the ebpf in the container cgroup? Maybe elimination may not be necessary due to inheritance from parents. In this case, even if it fails, the ebpf code for the container's cgroup device will remain, so there should be no such thing as the user being able to access all devices.

Reply:
@Furisto Since there is currently a way to get the contents of `config.json`, why don't we get the original information from here? To be honest, I want to get `status.json`, but `config.json` should be sufficient.
https://github.com/gitpod-io/gitpod/blob/60a9c550444f10dca15d4e39d12dc8d561500bc9/components/ws-daemon/pkg/container/containerd.go#L460-L470

Reply:
Hey @utam0k , that sounds like an optimization we could potentially add later, yes? In other words, let's try to keep this PR smaller, ship the skateboard, and ship another one in the future (if we need).

Aside from this observation, do you have any other concerns which are blocking? I ask so that we can get @Furisto the feedback needed to have this approved and merged.

Reply:
I did a little research and as it stands it shouldn't be a problem, because there is no API to change devices normally from k8s. However I think this should be improved. It can create some pretty strange bugs. The skateboard is a very good watchword 🛹